lib/Dialect/SV/Transforms/RTLLegalizeNames.cpp

```diff
@@ -0,0 +1,198 @@
+
+using namespace circt;
+using namespace sv;
+using namespace rtl;
+
+//===----------------------------------------------------------------------===//
+// RTLLegalizeNamesPass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct RTLLegalizeNamesPass
+    : public sv::RTLLegalizeNamesBase<RTLLegalizeNamesPass> {
+  void runOnOperation() override;
+
+private:
+  bool anythingChanged;
+
+  /// Given a module name \p moduleName, return the updated name if it has been
+  /// previously renamed in the table moduleNameTable, else call
+  /// sanitizeName, to get a new name in case of conflict with
+  /// keywords.
+  StringRef getModuleName(StringAttr moduleName) {
+    auto entryIter = moduleNameTable.find(moduleName);
+    if (entryIter != moduleNameTable.end())
+      return entryIter->getSecond();
+    auto updatedName = sanitizeName(moduleName.getValue(), usedModuleNames,
+                                    nextGeneratedNameID);
+    moduleNameTable[moduleName] = updatedName;
+    return updatedName;
+  }
+
+  /// moduleNameTable keeps track of mappings from Module names to updated names
+  /// to resolve keyword conflicts.
+  llvm::DenseMap<Attribute, StringRef> moduleNameTable;
+
+  /// Set of used MLIR module names, to ensure unique names when renaming
+  /// keywords in module names.
+  llvm::StringSet<> usedModuleNames;
+
+  size_t nextGeneratedNameID = 0;
+};
+} // end anonymous namespace
+
+void RTLLegalizeNamesPass::runOnOperation() {
+  anythingChanged = false;
+  auto root = getOperation();
+
+  // We run in two phases. Phase one legalizes module and interface names in the
+  // module globally. Phase two then dives into these structures and performs
+  // local legalization on names that cannot leak out their enclosing module.
+
+  // Register the names of external modules which we cannot rename.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto extMod = dyn_cast<RTLModuleExternOp>(op)) {
+      auto verilogName = extMod.getVerilogModuleNameAttr();
+      getModuleName(verilogName);
+      // We expressely *do not* rename extern modules, where the user supplies
+      // the complementary definition from the outside.
+      // TODO: Maybe we should emit a diagnostic if this pass would rename an
+      // external module.
+    }
+  });
+
+  // Rename modules and interfaces.
+  mlir::SymbolTableCollection symbolTable;
+  mlir::SymbolUserMap symbolUsers(symbolTable, root);
+  root.getBodyRegion().walk([&](Operation *op) {
+    StringAttr nameAttr;
+    if (auto module = dyn_cast<RTLModuleOp>(op)) {
+      nameAttr = module.getNameAttr();
+    } else if (auto intf = dyn_cast<InterfaceOp>(op)) {
+      nameAttr = intf.getNameAttr();
+    }
+    if (nameAttr) {
+      auto oldName = nameAttr.getValue();
+      auto newName = getModuleName(nameAttr);
+      if (oldName != newName) {
+        symbolUsers.replaceAllUsesWith(op, newName);
```

Comment:
I'm pretty sure this scans the whole IR, which will result in N^2 behavior. Am I correct @lattner ?

Reply:
If so, you should track the symbol renames and act on them in the walk() below.

Reply:
Upon construction the symbol table creates a `symbolToUsers` map (through `SymbolUserMap`) which allows for quick lookup of the users. That is a dense map so should be around O(log N) or close to O(1) for lookups, depending on the implementation.